Implement permission policies in the API

rijkvanzanten opened the pull request:
## Scope

What's changed:

- Implements policy-system based permissions handling on the API
- Replaces AuthorizationService with new set of functions in the `api/src/permissions` folder
- Allows roles to be nested
- Adds new `roles` flag to accountability object. This is an ordered array of all the parent roles of the current user
- Cleans up `get-ast-from-query` by splitting it into multiple files
- Permissions are now injected into the AST through `cases` and `whenCase`. This allows us to dynamically generate the case/when SQL to have dynamic field output per item.
- Cleans up `run-ast` by splitting it up into smaller files

## Potential Risks / Drawbacks

- The risks are very high. This replaces the full permissions system, and thus needs a lot of testing

## Review Notes / Questions

- This PR now compiles, but doesn't run yet.
- There was some weird logic happening in the users controller for TFA enable/disable that I'm not sure we need to keep. Needs a bit more testing 

## Todos

- [x] Add permissions processing for $CURRENT_USER etc flags
  - [x] Introduce $CURRENT_POLICIES and $CURRENT_ROLES for permissions
  - [ ] Decide if $CURRENT_POLICIES and $CURRENT_ROLES should be available in presets as well
- [ ] Add permissions merging when you're accessing from a share
- [x] Add caching to:
  - [x] Fetching Policies
  - [x] Fetching permissions
  - [x] Fetching the roles tree
  - [x] Fetching the field map
  - [x] Fetching allowed fields 
  - [x] Fetching allowed collections
- [x] Enable validation for admin users for wrong paths
- [x] Figure out what we want to do with Presets
- [x] Use `whenCases` in `run-ast` 
- [ ] Use applyCases in Meta Service for permissions aware counts
- [x] Handle admin-checks in roles and users services[^1] -> Handled in #22737 
- [x] Add unit testing for `clear` method in memory/cache
- [x] Make sure graphql gracefully handles optional fields when you have different fields for the same collection in multiple permission sets
- [ ] Add changeset
- [x] Unbork `/permissions` endpoint
- [x] Invalidate cache on permission changes
  - [x] On `directus_access` changes
  - [x] On `directus_roles` changes
  - [x] On `directus_permissions` changes
  - [x] On `directus_policies` changes
- [ ] Purge request level cache on permission related changes 
- [ ] Bring back check-ip middlewares?
- [ ] Fix `down` migration
- [x] Fix telemetry user counting -> Handled in #22737 
- [x] Fix user limit checking -> Handled in #22737 
- [ ] Reduce accountability used in the `withCache` to the known keys -> Handled in #22727
- [x] Fix `$FOLLOW` field key for filters (Fixed in https://github.com/directus/directus/pull/22384/commits/c00ff5df776d44073fceac49d88966bf74950edd)
- [ ] Fix handling of o2m fields that have partial permissions set (currently if some access to a o2m related field is given, it is returned for all items, even if the user should not see the field for some items)
- [ ] Fix explicit requests of o2m fields where the user does not have access to the related collection. (Right now this is silently dropped when generating the AST in order to avoid errors when requesting `*`).
- [ ] Account for user IP in the global request cache 
- [ ] Make sure that the globalAccess for a public user always has `app_access` and `admin_access` `false`
- [ ] Add new collections to GraphQL
- [ ] Add new collections to websockets
- [ ] Add new endpoints + types to SDK
- [x] Before merging this one in main, either retarget #22654 to main, or merge it in here

---

Closes #21778, closes #21765, closes #22163, closes #21769, closes #21768, closes #21767, closes #21766

[^1]: Eg check to make sure there's still >=1 admin left after the mutation is done

rijkvanzanten committed:
Update system data structure

rijkvanzanten committed:
Update migrations to keep structure flat

rijkvanzanten committed:
Remove icon from policies

rijkvanzanten committed:
Drop policies table on downgrade

rijkvanzanten committed:
Rearchitect migrations to structure v3

rijkvanzanten committed:
Add down migration

rijkvanzanten committed:
Update system fields

rijkvanzanten committed:
Add policy to fields import

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21760

paescuj committed:
Merge branch 'auditus' into auditus-21760

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21760

rijkvanzanten committed:
Fix public role attachment

rijkvanzanten committed:
Update packages/system-data/src/fields/index.ts

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

rijkvanzanten committed:
Update packages/system-data/src/fields/policies.yaml

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

rijkvanzanten committed:
Add nested roles

rijkvanzanten committed:
Remove unused step

rijkvanzanten committed:
Use two o2ms instead of m2a for attachments

rijkvanzanten committed:
Update system data

rijkvanzanten committed:
[WIP] Start reorging permissions handling

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21765

rijkvanzanten committed:
Setup field extraction

rijkvanzanten committed:
Remove watch from vitest

rijkvanzanten committed:
Finish fieldMap creation logic

rijkvanzanten committed:
Add tests for utils

rijkvanzanten committed:
Improve tests

rijkvanzanten committed:
Improve coverage

rijkvanzanten committed:
Split test and test:watch

rijkvanzanten committed:
Continue on this fun

rijkvanzanten committed:
[WIP] Setup processing

rijkvanzanten committed:
Sort roles

rijkvanzanten committed:
Restructure to util files for org

rijkvanzanten committed:
Add missing util tests

rijkvanzanten committed:
More tests

rijkvanzanten committed:
Add cases/whencase to ast

rijkvanzanten committed:
Start on injection logic

rijkvanzanten committed:
Add tests for inject cases

rijkvanzanten committed:
Add tests for process

rijkvanzanten committed:
Add todo

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21765

rijkvanzanten committed:
Organize run-ast

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21765

rijkvanzanten committed:
Add clear method to kv

rijkvanzanten committed:
Remove reliance on acc.perm

rijkvanzanten committed:
Restructure permissions setup

rijkvanzanten committed:
Drop perm from acc, add roles/policies

rijkvanzanten committed:
Remove get-permissions in middleware

rijkvanzanten committed:
Remove/comment use of acc.perm

rijkvanzanten committed:
Add default roles/permissions

rijkvanzanten committed:
Use knex

So we don't have to initialize the schema before we want to use the accountability system

rijkvanzanten committed:
Use new fetching logic in get accountability

rijkvanzanten committed:
Add new fetch global access utils

rijkvanzanten committed:
Gotta redo based on new setup

rijkvanzanten committed:
Replaced with new util

rijkvanzanten committed:
Remove dropping of perm in acc

It's no longer there by default, so no need to remove here

rijkvanzanten committed:
Temporarily comment out the enforce tfa check

rijkvanzanten committed:
Update usage of fetch tree to use knex

rijkvanzanten committed:
Don't store policies on accountability

rijkvanzanten committed:
Feed in roles thru acc

rijkvanzanten committed:
Bit of whitespace

rijkvanzanten committed:
Rename role->policy

rijkvanzanten committed:
Wreck some more stuff

Jk, this is splitting up the large get-ast-from-query function into smaller individual functions to make it easier to update the wildcard conversion to use permissions

rijkvanzanten committed:
Add ability to lookup all allowed fields in col+ac

rijkvanzanten committed:
Add note so I don't forget stuff which i will

rijkvanzanten committed:
Handle null acc

rijkvanzanten committed:
Introduce parseAst to itemsservice

rijkvanzanten committed:
That cleans things up

rijkvanzanten committed:
Replace checkAccess with validateAccess

rijkvanzanten committed:
Remove checkaccess from service

rijkvanzanten committed:
cleanup imports

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21765

rijkvanzanten committed:
Whoops one more

rijkvanzanten committed:
Leave crumbs for next time

rijkvanzanten committed:
Implement most of the fn

rijkvanzanten committed:
Fix various tests

rijkvanzanten committed:
Start on test for fetch roles tree

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21765

rijkvanzanten committed:
Add tests for fetch roles tree

rijkvanzanten committed:
Fix process tests

rijkvanzanten committed:
All. of. the. tests.

rijkvanzanten committed:
Update uses of validateAccess

rijkvanzanten committed:
Fix name in runAst

rijkvanzanten committed:
Fix use of accountability in gql sub

rijkvanzanten committed:
Deprecate authorization service

rijkvanzanten committed:
Remove getPermissions use

rijkvanzanten committed:
Drop old getpermissions

rijkvanzanten committed:
Pass services

rijkvanzanten committed:
Replace admin/app uses with fetch global

rijkvanzanten committed:
Update fetch user count to pull from policies

rijkvanzanten committed:
Remove broken admin existence checks

rijkvanzanten committed:
Update min accountability

rijkvanzanten committed:
Remove unused import

rijkvanzanten committed:
Drop permissions override from controller

rijkvanzanten committed:
Refactor reliance on acc.perm

rijkvanzanten committed:
Replace usage of permissions in fields

rijkvanzanten committed:
Replace usage of permissions in import/export

rijkvanzanten committed:
Drop permissions use from relations

rijkvanzanten committed:
Drop no longer used method

rijkvanzanten committed:
Remove unused import

rijkvanzanten committed:
fix type usage of pk in validate

rijkvanzanten committed:
Fix default acc in user

rijkvanzanten committed:
Replace use of permissions in utils

rijkvanzanten committed:
Update reduceSchema in specs/gql

rijkvanzanten committed:
Remove old share merging

rijkvanzanten committed:
Remove empty file

rijkvanzanten commented:
@DanielBiegler @hanneskuettner It'd be great if I could get a code-review while wrapping up the last todos. Once this thing is done and runs as it should, we'll involve the whole team for a good while of QA 🙂

rijkvanzanten committed:
Remove outdated comment

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21765

rijkvanzanten committed:
Use ctx objects for large param fns

hanneskuettner commented on api/src/auth/drivers/local.ts:
```diff
@@ -65,6 +65,10 @@ export function createLocalAuthRouter(provider: string): Router {
 			const accountability: Accountability = {
 				ip: getIPFromReq(req),
 				role: null,
+				user: null,
+				roles: [],
+				admin: false,
+				app: false,
```
Seeing this change (the addition of more properties to the Accountability context) in quite a few places. Would it make sense to abstract this away into a `createDefaultAccountability(ip?: string)` util, just so we have one place to change if we require more default parameters at a later point in time?

rijkvanzanten committed:
Add with-cache memoize util

rijkvanzanten committed:
Add cache to fetchpermissions

rijkvanzanten committed:
Update caching use in fetchRolesTree

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21765

rijkvanzanten committed:
Add caching to fetchAllowedFieldMap

rijkvanzanten committed:
Add more cache

rijkvanzanten committed:
Refactor call signatures

rijkvanzanten committed:
Move call signature updates

rijkvanzanten committed:
Handle presets

rijkvanzanten committed:
Update process call sig

rijkvanzanten committed:
Prevent infinite recursion in roles tree lookup

rijkvanzanten committed:
Use create util for acc

DanielBiegler reviewed (dismissed):
> [!NOTE]  
> TODO: We gotta update `/api/src/middleware/check-ip.ts` and use policies there.

Previously we got the `ip_access` of that users role and compared it with their IP. Done.

Now users could have multiple policies which each have an `ip_access` column.

rijkvanzanten commented:
@DanielBiegler Policies are filtered down by IP in the `fetch-policies` function, so every time you load in the policies set for the current accountability it'll be filtered by IP 🙂 

It's no longer a yes-or-no check in middleware; it's now filtering down the permissions for the specific request.

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21765

rijkvanzanten committed:
Remove old checkIp

rijkvanzanten committed:
Fix where equality operator

rijkvanzanten committed:
Break EVERYTHING!

Jk just cleaning up the structure some more, and removing the dep injection in favor of mocking

rijkvanzanten committed:
Fix build

rijkvanzanten committed:
Add missing module tests

rijkvanzanten committed:
Don't crash on missing parent

rijkvanzanten committed:
Fix role lookup

DanielBiegler committed:
add missing type annotation

DanielBiegler committed:
use logical-OR assignment and avoid a memory allocation

rijkvanzanten committed:
Attach admin policy in default admin creation

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21765

rijkvanzanten committed:
Fix admin check

rijkvanzanten committed:
Add todo for later

DanielBiegler committed:
rm code duplication

DanielBiegler committed:
fix test

it was missing the new `roles`

DanielBiegler committed:
add types and fix type error

policies dont [yet] have an icon

DanielBiegler committed:
move spread order to avoid potential future mishaps

new default keys would override the manually set keys, potentially leading to unintended behavior

DanielBiegler committed:
reduce allocations, add escape hatch to loop and type db-row

rijkvanzanten committed:
Implement case/when

rijkvanzanten committed:
Clean up comments

rijkvanzanten committed:
Optimize perm fetching in allowed f

rijkvanzanten committed:
Move apply case when to util fn

rijkvanzanten committed:
Optimize fetch-allowed-fields

rijkvanzanten committed:
Add fetch inconsistent util

rijkvanzanten committed:
Allow nulls

hanneskuettner committed:
Remove obsolete getCacheKey

hanneskuettner committed:
Remove unused import

hanneskuettner committed:
Update getAccountabilityForRole test

hanneskuettner committed:
Update fetchGlobalAccess test with one more test case + fix other test case

hanneskuettner committed:
Type cleanup

hanneskuettner committed:
Fix "admin access means automatic app access" in fetchGlobalAccessForQuery

hanneskuettner committed:
Clean up and expand fetch-inconsistent-field-map.test.ts

hanneskuettner commented on api/src/permissions/modules/process-ast/lib/inject-cases.ts:
```diff
@@ -0,0 +1,87 @@
+
+	const requestedKeys = children.map(({ fieldKey }) => fieldKey);
+
+	for (const [index, { rule, fields }] of rules.entries()) {
+		// If none of the fields in the current permissions rule overlap with the actually requested
+		// fields in the AST, we can ignore this case altogether
+		if (fields.has('*') === false && Array.from(fields).every((field) => requestedKeys.includes(field) === false)) {
+			continue;
+		}
+
+		cases.push(rule);
+
+		for (const field of fields) {
+			caseMap[field] = [...(caseMap[field] ?? []), index];
+		}
+	}
```
What does it mean if `rule` is `null` here? As far as I can tell it means that this permission does not have a permission filter, so it matches the `globalWhenCase` filter?  So effectively the question is what is the difference between `Permission.permissions == null` and `== {}`.

Right now `processChildren` returns `(Filter|null)[]`, which is an invalid type for `AST.cases`. 

So we should only add it to cases, if `rule != null` and use an independently incrementing index instead of the loop index?

```suggestion
	let index = 0;
	
	for (const { rule, fields } of rules) {
		// If none of the fields in the current permissions rule overlap with the actually requested
		// fields in the AST, we can ignore this case altogether
		if (fields.has('*') === false && Array.from(fields).every((field) => requestedKeys.includes(field) === false)) {
			continue;
		}

		if (rule === null) continue;
		
		cases.push(rule);

		for (const field of fields) {
			caseMap[field] = [...(caseMap[field] ?? []), index];
		}
		
		index++;
	}
```

rijkvanzanten replied:
> So effectively another question is what is the difference between Permission.permissions == null and == {}.

There's no difference between these two. `null` and `{}` should be treated the same in the case of the permission rules

rijkvanzanten replied:
> So we should only add it to cases, if rule != null and use an independently incrementing index instead of the loop index?

This is a good idea, as we can ignore any case that's null or `{}` anyways

hanneskuettner commented on api/src/permissions/modules/fetch-global-access/utils/fetch-global-access-for-query.ts:
```diff
@@ -0,0 +1,47 @@
+
+	// Additively merge access permissions
+	for (const { admin_access, app_access } of accessRows) {
+		globalAccess.admin ||= toBoolean(admin_access);
+		globalAccess.app ||= globalAccess.admin || toBoolean(app_access);
```
Hat to change this from Daniels change, in order for it to be compliant with the test (which I assume make the right assumption, that admin access means implied app access)

DanielBiegler replied:
Nice catch! :rocket:

hanneskuettner commented on api/src/utils/get-accountability-for-role.test.ts:
```diff
@@ -43,45 +42,50 @@ describe('getAccountabilityForRole', async () => {
+			admin: false,
+			app: false,
+			roles: [],
+			role: null,
```
Should this return `null` for `role`, the invalid role `'456-789'` or throw an error as it did before ?

rijkvanzanten commented on api/src/database/run/lib/get-db-query.ts:
```diff
@@ -0,0 +1,147 @@
+
+	// Queries with aggregates and groupBy will not have duplicate result
+	if (queryCopy.aggregate || queryCopy.group) {
+		const flatQuery = knex.from(table);
+		const query = await applyQuery(knex, table, flatQuery, queryCopy, schema, cases).query;
+		return query.select(fieldNodes.map(preProcess));
```
Order shouldn't matter. In this case it has to go through the `applyQuery` function; that's the change. Don't remember if there was a reason to feed it through `applyQuery` without the field selection yet..

rijkvanzanten commented on api/src/database/run/lib/parse-current-level.ts:
```diff
@@ -0,0 +1,68 @@
+	collection: string,
+	children: (NestedCollectionNode | FieldNode | FunctionFieldNode)[],
+	query: Query,
+	cases: Filter[],
```
Cases are used in run-ast now 🙂

hanneskuettner replied:
No, I mean the parameter is unused in this function, and could probably be removed or does it need to be incorporated here as well to also select additional columns that only appear in cases and are not extracted yet?

rijkvanzanten replied:
Ohhh my bad. Understand what you're saying now. They're no longer needed in this function. I thought I was gonna have to inject the cases here, but moved that elsewhere 🙂

rijkvanzanten commented on api/src/database/run/utils/apply-case-when.ts:
```diff
@@ -0,0 +1,48 @@
+): Knex.Raw {
+	const caseQuery = knex.queryBuilder();
+
+	applyFilter(knex, schema, caseQuery, { _or: columnCases }, table, aliasMap, cases);
```
That's the thinking, but we have to double check that 👍🏻

rijkvanzanten commented on api/src/permissions/lib/fetch-permissions.test.ts:
```diff
@@ -0,0 +1,50 @@
+import { beforeEach, expect, test, vi } from 'vitest';
+import { PermissionsService } from '../../services/permissions.js';
+import type { Context } from '../types.js';
+import { fetchPermissions as _fetchPermissions } from './fetch-permissions.js';
```
Oh that's a mix up and should be the other way around 👍🏻

hanneskuettner committed:
Test uncached functions

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21765

rijkvanzanten committed:
Test uncached

rijkvanzanten commented on api/src/permissions/modules/fetch-allowed-fields/fetch-allowed-fields.test.ts:
```diff
@@ -0,0 +1,44 @@
+import { PermissionsService } from '../../../services/permissions.js';
+import { fetchPolicies } from '../../lib/fetch-policies.js';
+import type { Context } from '../../types.js';
+import { fetchAllowedFields } from './fetch-allowed-fields.js';
```
Always without cache 👍🏻

rijkvanzanten committed:
Remove cases usage in parse-current-level

hanneskuettner committed:
Only consider non-null rules in inject cases

hanneskuettner committed:
Fix parseCurrentLevel call

hanneskuettner committed:
Move service imports into functions to avoid circular imports

hanneskuettner committed:
Ensure that we test that an error is thrown in processAst test

hanneskuettner committed:
Add failing test case for flattenFilter

hanneskuettner committed:
Ensure uniqueness in extractPathsFromQuery

hanneskuettner committed:
Early exit in validatePath

hanneskuettner committed:
Add additional test case for process payload test

hanneskuettner committed:
Update validateCollectionAccess test

hanneskuettner committed:
Clean up validate-item-access.test.ts

hanneskuettner committed:
Remove redundant initializer

hanneskuettner commented on api/src/services/items.ts:
```diff
@@ -419,31 +428,30 @@ export class ItemsService<Item extends AnyItem = AnyItem> implements AbstractSer
-			ast = await authorizationService.processAST(ast, opts?.permissionsAction);
-		}
+		ast = await processAst(
+			{ ast, action: 'read', accountability: this.accountability },
```
Did we drop support for `QueryOptions.permissionsAction` or is this accounted for somewhere else, that I'm missing?

hanneskuettner commented on api/src/services/meta.ts:
If we're planning to drop support for `meta` anyways (see [#15665](https://github.com/directus/directus/issues/15665)) I would suggest not investing any time in how to make this work and consider it a good point to drop it now rather than later?

hanneskuettner commented on api/src/services/relations.ts:
```diff
@@ -78,20 +94,26 @@ export class RelationsService {
-			if (permissions.fields.includes('*') === false) {
-				const allowedFields = permissions.fields;
-				if (allowedFields.includes(field) === false) throw new ForbiddenError();
+			if (allowedFields.includes(field) === false) {
```
This is missing the check for `'*'`. Was that done on purpose? As far as I can see `fetchAllowedFields` does not expand `'*'` to all available fields of a collection, so explicitly checking for `'*'` seems necessary?

hanneskuettner committed:
Use createDefaultAccountability

hanneskuettner committed:
Fix fetch-user-count.test.ts

hanneskuettner committed:
Cleanup unused default initializer

hanneskuettner committed:
Add empty cases to subfilter in _relationCount

hanneskuettner committed:
Drop AccessService and PermissionsService usage from services

hanneskuettner committed:
Found some more PermissionsServices

hanneskuettner committed:
Fix a few more tests

hanneskuettner committed:
Add nested role relation

hanneskuettner committed:
Fix query invocation in aggregate and group queries

hanneskuettner committed:
Fix role property name in auth/refresh

hanneskuettner commented on api/src/database/migrations/20240328A-permissions-policies.ts:
Does the `policy` fk in `directus_permissions` an `ON DELETE CASCADE` trigger? Otherwise deleting permissions does not work.

hanneskuettner replied:
Same goes for all fks in `directus_access` I guess?

rijkvanzanten replied:
Yes, _but_ we have to double check it all works as expected under MS SQL. It has some very strict limitations around circular triggers that could cause issues here

hanneskuettner commented on packages/types/src/policies.ts:
```diff
@@ -0,0 +1,9 @@
+	id: string;
+	name: string;
+	description: string;
+	enforce_tfa: null | boolean;
```
Do we want to include an `icon`, similar to what `roles` currently have? Make it look nice and pretty 😊

paescuj replied:
I wonder a bit whether it will be easy to find suitable/descriptive icons for policies 🤔 I'd probably leave it out of this PR, and then add it via the App PR if we see that it would actually be an enrichment.

rijkvanzanten replied:
Sure!

hanneskuettner committed:
Add some missing relations for permissions, access and roles

hanneskuettner committed:
Add m2o relation from permissions to policy

hanneskuettner committed:
Add m2o relation access to role, user, policy

hanneskuettner committed:
Allow fetchPermissions to fetch all permissions and not just those limited by an action

hanneskuettner committed:
Add parent to Role type

hanneskuettner commented on api/src/services/meta.ts:
```diff
@@ -38,16 +38,24 @@ export class MetaService {
+				},
+			);
+
+			// const policies = await fetchPolicies(this.accessService, this.accountability);
+			// const permissions = await fetchPermissions(this.permissionsService, 'read', policies, collection ? [collection] : undefined);
+
+			// TODO replace with applyCases (TBD)
+			// applyFilter(this.knex, this.schema, dbQuery, permissions, collection, {});
```
When we bring this back, we need to make sure to include `&& this.accountability.admin === true` in the if above, otherwise all fields will be filtered out.

2chiefk replied:
Yeah no kidding right

hanneskuettner commented on api/src/services/meta.ts:
```diff
@@ -58,27 +66,34 @@ export class MetaService {
+				},
+			);
+
+			// TODO rely on applyCases
+			// const permissions = permissionsRecord.permissions ?? {};
+
+			// if (Object.keys(filter).length > 0) {
+			// 	filter = { _and: [permissions, filter] };
+			// } else {
+			// 	filter = permissions;
+			// }
```
When we bring this back, we need to make sure to include `&& this.accountability.admin !== true` in the if above, otherwise all fields will be filtered out.

hanneskuettner committed:
Make sure that admin users see all fields

hanneskuettner committed:
Add access and policies controller, add util methods to policies and access service

hanneskuettner committed:
Change name and description of public policy, update description of admin policy and add on delete trigger.

hanneskuettner committed:
Make sure access row uuids are auto generated

DanielBiegler committed:
optimize kvredis clear function and add a unit test

to be fair: unit test is also testing implementation details but thats a problem there in general and for future us

hanneskuettner committed:
Add minimal app permission and dynamic variable injection to the permission fetching

hanneskuettner committed:
Fix m2o collection name in extractFieldsFromChildren

hanneskuettner committed:
Make sure to clone permission before injecting dynamic variables

hanneskuettner committed:
Actually do the cloning in with withAppMinimalPermissions since people might missbehave with the permissions obtained from PermissionsService.readByQuery so it better to go the source of the problem

hanneskuettner committed:
Use knex transaction in createOne -> processPayload - otherwise deadlock

hanneskuettner committed:
Make sure to respect '*' field in allowed fields

hanneskuettner committed:
Fix extractFieldsFromChildren for o2m as well - classic

hanneskuettner commented on api/src/controllers/policies.ts:
```diff
@@ -0,0 +1,220 @@
+router.search('/', validateBatch('read'), readHandler, respond);
+
+router.get(
+	'/me/flags',
```
Anyone have a better name for this? Since in the future it might not only be flags. Should it follow the naming of the libs and maybe be called `/me/global`? The idea behind this endpoint is to accumulate any properties that are derived when taking all policies of a user into account, like global access and currently the `enfore_tfa` flag.

DanielBiegler replied:
Mhh agreed `/me/globals` sounds better than flags to me too. 

> The idea behind this endpoint is to accumulate any properties

Mhh so maybe `/me/properties` ? :smile: 

_If_ we dont expand on it `/me/access` sounds pretty fitting for right now, but yeah that might be come less good in the future.

hanneskuettner replied:
Hmm, I personally dislike just `/me` because the endpoint does not return any policies, but some aggregation of all policies of the user, and we might end up with an actual `/me` endpoint that returns all policies at some point? wdyt?

paescuj replied:
Can you give an quick example of how the current response looks like?

hanneskuettner replied:
It is effectively a condensed version of whatever flags we have on policies right now, that apply to a user.

```json
{ "app_access": true, "admin_access": false, "enforce_tfa": false }
```

paescuj replied:
Okay, I think `global(s)` sounds more appropriate, that's ultimately what we're retrieving here: global flags/properties/whatever.

>  but some aggregation of all policies of the user, and we might end up with an actual `/me` endpoint that returns all policies at some point? wdyt?

You mean at a later point or is that something you'd like to look into now?

hanneskuettner replied:
At a later point. Currently I don't see the need of retrieving all the policies of a user efficiently, as the app does not need to know them to work. But maybe that need arises at some point in the future, so I'd like to keep that option alive.

hanneskuettner committed:
Fix allowed field check in `FieldsService.readAll` to account for multiple permissions for collection+action

hanneskuettner committed:
Skip case/when if `allowedFields` includes '*'

hanneskuettner committed:
Restructure the way the current users permissions are returned

DanielBiegler committed:
add ability to clear all keys from memory cache

DanielBiegler committed:
add test for clear method

DanielBiegler committed:
add await to clear function

hanneskuettner committed:
Clear permissions caches on changes to policy attachments (directus_access) and policy updates (directus_policies) and permissions updates (directus_permissions)

hanneskuettner committed:
Make the public role a real role rather than a virtual one

hanneskuettner commented on api/src/permissions/modules/process-ast/utils/validate-path.ts:
```diff
@@ -0,0 +1,48 @@
+	const permissionsForCollection = permissions.filter((permission) => permission.collection === collection);
+
+	if (permissionsForCollection.length === 0) {
+		throw new ForbiddenError({
+			reason: `You don't have permission to access collection "${collection}" or it does not exist. Queried in ${pathSuffix}.`,
+		});
```
Even though this, and the other errors are really helpful, they do expose that the collection + field exists, as it differs from the standard `ForbiddenError` in case of a non-existent collection.

rijkvanzanten replied:
This error returns the collections/fields that the user already provided in their query, so it shouldn't return any new information they didn't already provide. It merely shows you where in the query the wrong collection/field is, but doesn't expose whether that collection/field is real

hanneskuettner committed:
Inject the public role, we're it previously was `null`

hanneskuettner committed:
Revert adding a fix public role

DanielBiegler committed:
remove unused variable

hanneskuettner committed:
Ensure that a user without a role can still use the /me util endpoints

hanneskuettner committed:
Make sure that the /me endpoints always return minimal information, similar to /users/me

hanneskuettner committed:
Merge remote-tracking branch 'refs/remotes/origin/auditus' into auditus-21765

# Conflicts:
#	api/src/services/users.ts
#	api/src/telemetry/lib/get-report.test.ts
#	api/src/telemetry/lib/get-report.ts
#	api/src/telemetry/utils/get-user-count.test.ts
#	api/src/telemetry/utils/get-user-count.ts

hanneskuettner committed:
Merge branch 'refs/heads/auditus' into auditus-21765

# Conflicts:
#	api/src/services/users.ts

licitdev commented on api/src/controllers/roles.ts:
```diff
@@ -73,6 +73,36 @@ const readHandler = asyncHandler(async (req, res, next) => {
+			schema: req.schema,
+		});
+
+		const query = { ...req.sanitizedQuery, limit: -1 };
+
+		try {
+			const roles = await service.readMany(req.accountability.roles, query);
```
Should we include this comment about it being temporary to be consistent with other usages?

```suggestion
		// TODO fix this at the service level
		const temporaryQuery = { ...req.sanitizedQuery, limit: -1 };

		try {
			const roles = await service.readMany(req.accountability.roles, temporaryQuery);
```

hanneskuettner committed:
Some fixes after merging main

rijkvanzanten commented on api/src/database/run/lib/get-db-query.ts:
```diff
@@ -0,0 +1,144 @@
+): Promise<Knex.QueryBuilder> {
+	const env = useEnv();
+	const preProcess = getColumnPreprocessor(knex, schema, table);
+	const queryCopy = clone(query);
```
Probably an accident. We should prevent any changes to the original query object

rijkvanzanten commented on api/src/permissions/utils/filter-policies-by-ip.test.ts:
```diff
@@ -0,0 +1,59 @@
+	]);
+});
+
+test('Keeps policies that match the IP cidr block', () => {
```
I'm always a little torn between doing tests purely unit-style or more integration-style. In this case, I don't care what filter-policies-by-ip does under the hood, as long as the result is correct. That does indeed mean we're double-testing that ipInNetworks function in here as well

rijkvanzanten committed:
Update api/src/permissions/utils/with-cache.ts

Co-authored-by: Hannes Küttner <4376726+hanneskuettner@users.noreply.github.com>

licitdev commented on api/src/database/run/run.ts:
Should this file be `run-ast.ts`?

hanneskuettner committed:
Avoid broken role query for now

hanneskuettner committed:
Merge remote-tracking branch 'origin/auditus-21765' into auditus-21765

licitdev commented on api/src/permissions/modules/process-ast/process.ts:
```diff
@@ -0,0 +1,45 @@
+	);
+
+	for (const [path, { collection, fields }] of fieldMap.entries()) {
+		validatePath(path, permissions, collection, fields);
```
`validatePath` errors when querying with the default `fields=*` when there is no permission for the related collection in relational fields.

The previous behaviour "ignored" the relational field without permission, by not including it within the AST.

Previous logic that is now in `parse-fields.ts` with the permission check removed: 
https://github.com/directus/directus/blob/9e91405677bb1803c6159605914b2015bd2cdcee/api/src/utils/get-ast-from-query.ts#L246-L249

hanneskuettner replied:
Fixed in https://github.com/directus/directus/pull/22384/commits/3993ea5a06533ae41dfb8712f9adf4672f19c30f

hanneskuettner committed:
Skip related collection `parseFields` if user has no permissions

hanneskuettner committed:
Ensure same call order as in `convertWildcards`

hanneskuettner committed:
Create default admin policy and connect it in cli init command

hanneskuettner commented on api/src/cli/commands/init/index.ts:
```diff
@@ -101,10 +101,7 @@ export default async function init(): Promise<void> {
-		id: roleID,
-		...defaultAdminRole,
-	});
+	await db('directus_roles').insert({ ...defaultAdminRole, id: roleID });
```
Fixed in https://github.com/directus/directus/pull/22384/commits/6507b129204be52864114a7d1dc518e8b912021d

hanneskuettner committed:
Remove obsolete middleware mock in app.test.ts

hanneskuettner commented on api/src/app.test.ts:
```diff
@@ -47,10 +47,6 @@ vi.mock('./flows', () => ({
 	}),
 }));
 
-vi.mock('./middleware/check-ip', () => ({
-	checkIP: Router(),
-}));
-
```
Removed in https://github.com/directus/directus/pull/22384/commits/f3b7e7e7ffd9ebf866e559b1511c7602be08d162

hanneskuettner committed:
Add validation against non-existent fields and collections to `validatePath`

hanneskuettner committed:
Split up permission and path existence validation and validate path existence for admin users as well

hanneskuettner committed:
Make applySearch not async

hanneskuettner committed:
Fix relation extraction and permissions for `$FOLLOW` fields

hanneskuettner committed:
Fix case when for related collections and query wrapping

hanneskuettner committed:
Merge branch 'auditus' into auditus-21765

hanneskuettner committed:
Rework user integrity checks for Auditus (#22737)

* Changes to user counting and integrity checks

* Ensure that user validation happens in both create one and create many

* Rename `checkType` to `flags`

* Update api/src/permissions/modules/validate-remaining-admin/validate-remaining-admin-count.ts

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

* Update to enum usage

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>

* A few more changes to enum instead of number

* One more enum type update

* Make sure to correctly override the callback when combining options

* Clean up option type

* Update api/src/services/users.ts

Co-authored-by: ian <licitdev@gmail.com>

* Only take validation shortcut for users

We can only be sure that the deletion of users does not increase any other access types count, so in all other cases we need to verify that for example the App or API users have not increased over the limit

* Make both app and admin users count against app access limit

* Update api/src/permissions/modules/validate-remaining-admin/validate-remaining-admin-count.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

* One post-merge fix, two small fixes

* Simplify flag updating and callback calling

* Changing app access in a policy only requires user limit checking, not full check

* Only the status of a created user should matter to determine if a check is neccessary

* Add count alias to count query

---------

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>
Co-authored-by: ian <licitdev@gmail.com>
Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>
Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>

hanneskuettner committed:
Add roles and permissions to the app (#22654)

* Initial app changes

* Fix getRelationsForField

* Add changeset

* Remove app-permissions from role settings

* Make sure access row uuids are auto generated

* Move a few things around, set up policies m2m properly

* Show roles as tree in sidebar
Change avatar field query for user

* Show user and role count in policy table

* Default to not adding app access for a policy, makes composability less annoying

* Correctly fall back to 0 for counts

* Change the structure of current user permissions

* Start bringing back the public role

* Make the public role a real role rather than a virtual one

* Revert public role changes

* Extend list-m2m to allow for very custom junction matching and a primary key of `null`

* Remove unused

* Fix public role policy update payload

* Fix app access for users without role (which is a thing now apparently)

* Make sure that the /me endpoints always return minimal information, similar to /users/me

* Tweak nav icons

* Pull policy id from constants

* Update permissions interface design to match

New design language in figma

* Some minor adjustments

- Make chip hover border more consistent
- Add "Remove" button to remove a full row of permissions, as in the UI mockup
- Fix table layout

* Clean up a few more things

* Fix `setFullAccess`

* Align collection view icons with navigation

* Don't query 'admin_access' for role

* Fix relation extraction and permissions for `$FOLLOW` fields

* Don't show `0 Items` for child rows, but `--` instead

* Make policy detail work in nested policy creating use case

* Remove unused v-icon override

* Move system collections to separate visual table

* Navigate before refresh

Prevents a flash of the previous value to be visible in the table

* Move composable to separate file

---------

Co-authored-by: Daniel Biegler <DanielBiegler@users.noreply.github.com>
Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>

rijkvanzanten committed:
Merge branch 'auditus' into auditus-21765

rijkvanzanten committed:
Optimize types

rijkvanzanten committed:
Clone query deep

rijkvanzanten committed:
Optimize type order

rijkvanzanten commented on api/src/permissions/modules/fetch-global-access/lib/fetch-global-access-for-roles.ts:
```diff
@@ -0,0 +1,9 @@
+import type { GlobalAccess } from '../types.js';
+import { fetchGlobalAccessForQuery } from '../utils/fetch-global-access-for-query.js';
+
+export async function fetchGlobalAccessForRoles(knex: Knex, roles: string[]): Promise<GlobalAccess> {
+	const query = knex.where('role', 'in', roles);
+	return await fetchGlobalAccessForQuery(query, `global-access-roles-${getSimpleHash(JSON.stringify(roles))}`);
+}
```
Hm yes no maybe. The `Knex` object passed doesn't serialize down that neatly, so I'm a tad worried about the hashing performance with such a large query object

hanneskuettner replied:
Oh I would definitely restructure that into `(options, context)` first, just like all the other functions that use with cache

hanneskuettner replied:
No need to hash the query object if we just look at the user / roles for the cache key anyways

rijkvanzanten committed:
Throw error on invalid role id

rijkvanzanten committed:
Rename run.js -> run-ast.js

hanneskuettner committed:
Re-add filesizes to telemetry report collection that got lost in the merge

hanneskuettner committed:
Make `systemCollections` reactive

hanneskuettner committed:
Use one column per action to avoid unwanted shifting if some actions are not allowed at all

rijkvanzanten committed:
Render system and custom together

hanneskuettner committed:
Add divider between regular and system permissions if both have elements

licitdev commented on api/src/permissions/modules/process-ast/lib/field-map-from-ast.ts:
```diff
@@ -0,0 +1,14 @@
+export function fieldMapFromAst(ast: AST, schema: SchemaOverview): FieldMap {
+	const fieldMap: FieldMap = new Map();
+
+	extractFieldsFromChildren(ast.name, ast.children, fieldMap, schema);
+	extractFieldsFromQuery(ast.name, ast.query, fieldMap, schema);
```
Alias fields should be accounted for here as invalid fields are currently returned, causing errors when the path existence is validated. 🤔

hanneskuettner replied:
Fixed in https://github.com/directus/directus/pull/22384/commits/a1870749aa34808f19285d3423cf6016cc343b41

hanneskuettner committed:
Add AccessService and PoliciesService to `getService`

hanneskuettner committed:
Move policy global flags fetching to util

hanneskuettner committed:
Move collection access fetching into util

hanneskuettner committed:
Remove permissions for `directus_access`, `directus_permissions` and `directus_policies` from schema permissions

hanneskuettner committed:
Merge branch 'auditus' into auditus-21765

hanneskuettner committed:
use formatted-value display for name & description in roles & policies

hanneskuettner committed:
Rename `process.ts` to `process-ast.ts`

paescuj committed:
Fix process-ast import after renaming

hanneskuettner committed:
Perform user integrity check on item deletion

paescuj committed:
Fix first admin creation on bootstrap

paescuj committed:
Revert "Fix first admin creation on bootstrap"

This reverts commit bf480d023c508a93e6d7304ae8f52c70af47da7e.

Will be fixed by adjusting the check in access service

hanneskuettner committed:
Don't perform admin integrity check if a new access row is created. Only check user limits

hanneskuettner committed:
Don't set an alias to the raw column value if it is wrapped in a case/when

hanneskuettner committed:
Correctly handle aliases when in field map and case injection
